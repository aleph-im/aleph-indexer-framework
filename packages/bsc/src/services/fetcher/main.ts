import {
  Blockchain,
  BlockchainFetcherI,
  FetcherMsClient,
  IndexableEntityType,
  BaseEntityFetcherMain,
} from '@aleph-indexer/framework'
import { EthereumFetcher } from '@aleph-indexer/ethereum'
import { BscBlockHistoryFetcher } from './src/block/blockHistoryFetcher.js'

export class BscFetcher extends EthereumFetcher implements BlockchainFetcherI {
  constructor(
    protected fetcherClient: FetcherMsClient,
    protected blockHistoryFetcher: BscBlockHistoryFetcher,
    protected entityFetchers: Partial<
      Record<IndexableEntityType, BaseEntityFetcherMain<any, any, any>>
    >,
    protected blockchainId: Blockchain = Blockchain.Bsc,
  ) {
    super(fetcherClient, blockHistoryFetcher, entityFetchers, blockchainId)
  }
}
