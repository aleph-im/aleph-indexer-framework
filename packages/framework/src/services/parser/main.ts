import { ServiceBroker } from 'moleculer'
import {
  ParsedTransactionV1,
  ParsedInstructionV1,
  ParsedAccountInfoV1,
  RawTransaction,
  RawInstruction,
  RawAccountInfo,
} from '@aleph-indexer/core'
import { ParserMsI } from './interface.js'
import { TransactionParser } from './src/transactionParser.js'
import { AccountParserLibrary } from './src/accountParserLibrary.js'
import { InstructionParserLibrary } from './src/instructionParserLibrary.js'
import { MsIds, MsMainWithEvents } from '../common.js'

export class ParserMsMain extends MsMainWithEvents implements ParserMsI {
  constructor(
    protected broker: ServiceBroker,
    protected layoutPath?: string,
    protected instructionParserLibrary: InstructionParserLibrary = new InstructionParserLibrary(
      layoutPath,
    ),
    protected accountParserLibrary: AccountParserLibrary = new AccountParserLibrary(),
    protected transactionParser: TransactionParser = new TransactionParser(
      instructionParserLibrary,
    ),
  ) {
    super(broker, MsIds.Parser)
  }

  async onTxs(chunk: RawTransaction[]): Promise<void> {
    // console.log(`📩 ${chunk.length} txs received by the parser...`)

    const parsedTxs: ParsedTransactionV1[] = []

    for (const tx of chunk) {
      try {
        const result = await this.parseTransaction(tx)
        parsedTxs.push(result)
      } catch (e) {
        console.error(e)
        continue
      }
    }

    await this.emitTransactions(parsedTxs)
  }

  async parseTransaction(
    payload: RawTransaction,
  ): Promise<ParsedTransactionV1> {
    return this.transactionParser.parse(payload)
  }

  async parseInstruction(
    payload: RawInstruction,
  ): Promise<RawInstruction | ParsedInstructionV1> {
    return await this.instructionParserLibrary.parse(payload)
  }

  async parseAccountData(
    account: string,
    payload: RawAccountInfo,
  ): Promise<RawAccountInfo | ParsedAccountInfoV1> {
    return await this.accountParserLibrary.parse(payload, account)
  }

  protected async emitTransactions(txs: ParsedTransactionV1[]): Promise<void> {
    if (!txs.length) return

    console.log(`✉️  ${txs.length} txs sent by the parser...`)

    const txGroups = this.groupTransactions(txs)
    const groups = this.getClientEventGroups()

    await Promise.all(
      groups.flatMap((group) =>
        Object.entries(txGroups).map(([partitionKey, txs]) =>
          this.emitToClients('txs', txs, { group, partitionKey }),
        ),
      ),
    )

    // return this.broadcastToClients('txs', txs)
    // return this.broker.broadcast('parser.txs', txs, [MsIds.Indexer])
  }

  protected groupTransactions(
    txs: ParsedTransactionV1[],
  ): Record<string, ParsedTransactionV1[]> {
    return txs.reduce((acc, tx) => {
      tx.parsed.message.accountKeys.forEach(({ pubkey }) => {
        const byAccount = acc[pubkey] || (acc[pubkey] = [])
        byAccount.push(tx)
      })
      return acc
    }, {} as Record<string, ParsedTransactionV1[]>)
  }
}
