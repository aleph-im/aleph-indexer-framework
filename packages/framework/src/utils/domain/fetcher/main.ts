import { Utils } from '@aleph-indexer/core'
import {
  FetcherMainDomainContext,
  FetcherState,
  SignatureFetcherState,
  TransactionState,
} from '../../../services/fetcher/src/types.js'

/**
 * The main fetcher domain class.
 * Primary entry point for information about fetcher state for the indexer.
 */
export class FetcherMainDomain {
  protected discoverJob: Utils.JobRunner | undefined
  protected statsJob: Utils.JobRunner | undefined
  protected accounts: Set<string> = new Set()

  constructor(protected context: FetcherMainDomainContext) {}

  /**
   * Returns the fetcher state for the given account-bound fetchers.
   * @param accounts The accounts of which to get the fetchers.
   */
  async getAccountFetcherState(
    accounts: string[] = [],
  ): Promise<SignatureFetcherState[]> {
    return (
      await Promise.all(
        accounts.map((account) =>
          this.context.apiClient.getAccountFetcherState({
            account,
          }),
        ),
      )
    ).filter((info): info is SignatureFetcherState => !!info)
  }

  /**
   * Returns the fetcher state for the given non-account related fetchers.
   * @param fetchers The fetchers to get the fetcher state for.
   */
  async getFetcherState(fetchers: string[] = []): Promise<FetcherState[]> {
    fetchers = fetchers.length
      ? fetchers
      : this.context.apiClient.getAllFetchers()

    return Promise.all(
      fetchers.map((fetcher) =>
        this.context.apiClient.getFetcherState({
          fetcher,
        }),
      ),
    )
  }

  /**
   * Returns the state of being fetched for the given transactions.
   * @param signatures The signatures of the transactions to get the state for.
   */
  async getTransactionState(
    signatures: string[] = [],
  ): Promise<TransactionState[]> {
    return this.context.apiClient.getTransactionState({
      signatures,
    })
  }
}
