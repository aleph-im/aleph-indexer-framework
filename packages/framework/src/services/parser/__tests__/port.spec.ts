import { jest } from '@jest/globals'

jest.useFakeTimers()

import { ParserMsMain } from '../main.js'
import { port1 } from '../__mocks__/txs/port.js'

describe('ParserMs integration', () => {
  let parserMs: ParserMsMain

  beforeEach(() => {
    const mockObj: any = jest.fn()
    parserMs = new ParserMsMain(mockObj, mockObj, mockObj)
  })

  it('port raw transaction should be parsed', async () => {
    const parsedTx = await parserMs.parseTransaction(port1)
    expect(parsedTx).toMatchSnapshot()
  })
})
