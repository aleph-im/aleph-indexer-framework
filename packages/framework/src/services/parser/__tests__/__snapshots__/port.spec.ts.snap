// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParserMs integration port raw transaction should be parsed 1`] = `
Object {
  "blockTime": 1630046289,
  "meta": Object {
    "err": null,
    "fee": 5000,
    "innerInstructions": Array [],
    "loadedAddresses": Object {
      "readonly": Array [],
      "writable": Array [],
    },
    "logMessages": Array [
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Reserve",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3375 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Reserve",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 5313 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Reserve",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 5313 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Reserve",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 5313 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Reserve",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 5313 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Reserve",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 5313 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Reserve",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3375 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15271 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3162 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15222 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3162 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15271 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3162 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15294 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15227 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3162 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3162 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15271 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3162 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 3162 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15281 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR invoke [1]",
      "Program log: Instruction: Refresh Obligation",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR consumed 15227 of 200000 compute units",
      "Program Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR success",
    ],
    "postBalances": Array [
      339943850553,
      4892880,
      4892880,
      4892880,
      4892880,
      4892880,
      4892880,
      4892880,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      1,
      23942400,
      23942400,
      23942400,
      23942400,
      23942400,
      1141440,
    ],
    "postTokenBalances": Array [],
    "preBalances": Array [
      339943855553,
      4892880,
      4892880,
      4892880,
      4892880,
      4892880,
      4892880,
      4892880,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      7266240,
      1,
      23942400,
      23942400,
      23942400,
      23942400,
      23942400,
      1141440,
    ],
    "preTokenBalances": Array [],
    "rewards": Array [],
    "status": Object {
      "Ok": null,
    },
  },
  "parsed": Object {
    "message": Object {
      "accountKeys": Array [
        Object {
          "pubkey": "EsQ179Q8ESroBnnmTDmWEV4rZLkRc3yck32PqMxypE5z",
          "signer": true,
          "writable": true,
        },
        Object {
          "pubkey": "DSw99gXoGzvc4N7cNGU7TJ9bCWFq96NU2Cczi1TabDx2",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "DSST29PMCVkxo8cf5ht9LxrPoMc8jAZt98t6nuJywz8p",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "ZgS3sv1tJAor2rbGMFLeJwxsEGDiHkcrR2ZaNHZUpyF",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "X9ByyhmtQH3Wjku9N5obPy54DbVjZV7Z99TPJZ2rwcs",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "BnhsmYVvNjXK3TGDHLj1Yr1jBGCmD1gZMkAyCwoXsHwt",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "4tqY9Hv7e8YhNQXuH75WKrZ7tTckbv2GfFVxmVcScW5s",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "DcENuKuYd6BWGhKfGr7eARxodqG12Bz1sN5WA8NwvLRx",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "GqGTh2BbWwYWiPJgFy83SEU5oxdFQmhokRbLAq7RwG4r",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "9jAKR48m4Q33qBfMfjX5aQNQcKkVznohEw99J8Ps8Fxu",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "2zCGoXwaXhcngx7VfFYP9hg6kqu4hYk61yAjjYwEiJhR",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "EGohXcPLcJ15z33GkXCFgZoFCdY2uz2CewZ7bac3Amex",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "9R2t7KxkrWa8wRgaKCf21s4uCZPDX8nwgQ2EzDg3RSqL",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "4dUB4CGj4oqgTeDedfhCy8VvshYpJcM4jytqyKCxUbTf",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "GGUan1549HYvpDWL4T96Hh3Fx3WE8ZmAdJxhXz45hdLT",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "AnDVjVxmb9TQQLCzs979v21LH66k32zY7Lzgg8Cko5Vw",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "DyEsnDYq6pyYnBm177BMaDUzggyjMiVcZVALmiBYaYhB",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "FLsN5kM8SXj5akRCreh2zPF82k8UZwbCjhj5fbW8Gux7",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "3GtnY1CTHU1TiqWEKnyp3y9NsjXsN13XMWVE1F6oG5h3",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "3qDNa1xPcBzYYJvJYS4EZddeSpLswMhtkse8r3HoRDqr",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "Dys1u8HWyoh7hVe8vN6JYCHD3Kt7uTtZebGaKNyzpGoo",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "HZznSGNNmDB5vjVXPYFTdAHXgqyF2y99b37B4Hyey6o6",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "6iPsoZshp72ZkfUWcVTvJJR4KfHQifuPuAXSKEvDZWTz",
          "signer": false,
          "writable": true,
        },
        Object {
          "pubkey": "SysvarC1ock11111111111111111111111111111111",
          "signer": false,
          "writable": false,
        },
        Object {
          "pubkey": "GVXRSBjFk6e6J3NbVPXohDJetcTjaeeuykUpbQF8UoMU",
          "signer": false,
          "writable": false,
        },
        Object {
          "pubkey": "3NBReDRTLKMQEKiLD5tGcx4kXbTf88b7f2xLS9UuGjym",
          "signer": false,
          "writable": false,
        },
        Object {
          "pubkey": "H6ARHf6YXhGYeQfUzQNGk6rDNnLBQKrenN712K4AQJEG",
          "signer": false,
          "writable": false,
        },
        Object {
          "pubkey": "G4AQpTYKH1Fmg38VpFQbv6uKYQMpRhJzNPALhp7hqdrs",
          "signer": false,
          "writable": false,
        },
        Object {
          "pubkey": "3vxLXJqLqF3JG5TCbYycbKWRBbCJQLxQmBGCkyqEEefL",
          "signer": false,
          "writable": false,
        },
        Object {
          "pubkey": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
          "signer": false,
          "writable": false,
        },
      ],
      "addressTableLookups": null,
      "instructions": Array [
        Object {
          "accounts": Array [
            "DSw99gXoGzvc4N7cNGU7TJ9bCWFq96NU2Cczi1TabDx2",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "4",
          "index": 0,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "DSST29PMCVkxo8cf5ht9LxrPoMc8jAZt98t6nuJywz8p",
            "SysvarC1ock11111111111111111111111111111111",
            "GVXRSBjFk6e6J3NbVPXohDJetcTjaeeuykUpbQF8UoMU",
          ],
          "data": "4",
          "index": 1,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "ZgS3sv1tJAor2rbGMFLeJwxsEGDiHkcrR2ZaNHZUpyF",
            "SysvarC1ock11111111111111111111111111111111",
            "3NBReDRTLKMQEKiLD5tGcx4kXbTf88b7f2xLS9UuGjym",
          ],
          "data": "4",
          "index": 2,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "X9ByyhmtQH3Wjku9N5obPy54DbVjZV7Z99TPJZ2rwcs",
            "SysvarC1ock11111111111111111111111111111111",
            "H6ARHf6YXhGYeQfUzQNGk6rDNnLBQKrenN712K4AQJEG",
          ],
          "data": "4",
          "index": 3,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "BnhsmYVvNjXK3TGDHLj1Yr1jBGCmD1gZMkAyCwoXsHwt",
            "SysvarC1ock11111111111111111111111111111111",
            "G4AQpTYKH1Fmg38VpFQbv6uKYQMpRhJzNPALhp7hqdrs",
          ],
          "data": "4",
          "index": 4,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "4tqY9Hv7e8YhNQXuH75WKrZ7tTckbv2GfFVxmVcScW5s",
            "SysvarC1ock11111111111111111111111111111111",
            "3vxLXJqLqF3JG5TCbYycbKWRBbCJQLxQmBGCkyqEEefL",
          ],
          "data": "4",
          "index": 5,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "DcENuKuYd6BWGhKfGr7eARxodqG12Bz1sN5WA8NwvLRx",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "4",
          "index": 6,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "GqGTh2BbWwYWiPJgFy83SEU5oxdFQmhokRbLAq7RwG4r",
            "SysvarC1ock11111111111111111111111111111111",
            "BnhsmYVvNjXK3TGDHLj1Yr1jBGCmD1gZMkAyCwoXsHwt",
          ],
          "data": "8",
          "index": 7,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "9jAKR48m4Q33qBfMfjX5aQNQcKkVznohEw99J8Ps8Fxu",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "8",
          "index": 8,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "2zCGoXwaXhcngx7VfFYP9hg6kqu4hYk61yAjjYwEiJhR",
            "SysvarC1ock11111111111111111111111111111111",
            "DcENuKuYd6BWGhKfGr7eARxodqG12Bz1sN5WA8NwvLRx",
          ],
          "data": "8",
          "index": 9,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "EGohXcPLcJ15z33GkXCFgZoFCdY2uz2CewZ7bac3Amex",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "8",
          "index": 10,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "9R2t7KxkrWa8wRgaKCf21s4uCZPDX8nwgQ2EzDg3RSqL",
            "SysvarC1ock11111111111111111111111111111111",
            "BnhsmYVvNjXK3TGDHLj1Yr1jBGCmD1gZMkAyCwoXsHwt",
          ],
          "data": "8",
          "index": 11,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "4dUB4CGj4oqgTeDedfhCy8VvshYpJcM4jytqyKCxUbTf",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "8",
          "index": 12,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "GGUan1549HYvpDWL4T96Hh3Fx3WE8ZmAdJxhXz45hdLT",
            "SysvarC1ock11111111111111111111111111111111",
            "BnhsmYVvNjXK3TGDHLj1Yr1jBGCmD1gZMkAyCwoXsHwt",
          ],
          "data": "8",
          "index": 13,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "AnDVjVxmb9TQQLCzs979v21LH66k32zY7Lzgg8Cko5Vw",
            "SysvarC1ock11111111111111111111111111111111",
            "DcENuKuYd6BWGhKfGr7eARxodqG12Bz1sN5WA8NwvLRx",
          ],
          "data": "8",
          "index": 14,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "DyEsnDYq6pyYnBm177BMaDUzggyjMiVcZVALmiBYaYhB",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "8",
          "index": 15,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "FLsN5kM8SXj5akRCreh2zPF82k8UZwbCjhj5fbW8Gux7",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "8",
          "index": 16,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "3GtnY1CTHU1TiqWEKnyp3y9NsjXsN13XMWVE1F6oG5h3",
            "SysvarC1ock11111111111111111111111111111111",
            "BnhsmYVvNjXK3TGDHLj1Yr1jBGCmD1gZMkAyCwoXsHwt",
          ],
          "data": "8",
          "index": 17,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "3qDNa1xPcBzYYJvJYS4EZddeSpLswMhtkse8r3HoRDqr",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "8",
          "index": 18,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "Dys1u8HWyoh7hVe8vN6JYCHD3Kt7uTtZebGaKNyzpGoo",
            "SysvarC1ock11111111111111111111111111111111",
          ],
          "data": "8",
          "index": 19,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "HZznSGNNmDB5vjVXPYFTdAHXgqyF2y99b37B4Hyey6o6",
            "SysvarC1ock11111111111111111111111111111111",
            "BnhsmYVvNjXK3TGDHLj1Yr1jBGCmD1gZMkAyCwoXsHwt",
          ],
          "data": "8",
          "index": 20,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
        Object {
          "accounts": Array [
            "6iPsoZshp72ZkfUWcVTvJJR4KfHQifuPuAXSKEvDZWTz",
            "SysvarC1ock11111111111111111111111111111111",
            "DcENuKuYd6BWGhKfGr7eARxodqG12Bz1sN5WA8NwvLRx",
          ],
          "data": "8",
          "index": 21,
          "programId": "Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR",
        },
      ],
      "recentBlockhash": "EDt142jiUUw5daz5uAnv6oBhkB2LZFJCqbfvbShbegBi",
    },
    "signatures": Array [
      "hQG8qhRhdmdAfBtAfLMjLMSRcCtaavAhdtj9ybSL5e2WfgmpfAAbx4JNMNFhPeXnH3PZ64ijq8kugTkTRDEvLyQ",
    ],
  },
  "signature": "hQG8qhRhdmdAfBtAfLMjLMSRcCtaavAhdtj9ybSL5e2WfgmpfAAbx4JNMNFhPeXnH3PZ64ijq8kugTkTRDEvLyQ",
  "slot": 93548810,
}
`;
